### DISCUSSION TITLE
TypeScript: Type inference question with TS v4.2

### DISCUSSION THREAD

**Discussion by oscard0m** [octokit/octokit.js#2023] [Ideas]
Hi all,

I tried to find out a conversation we had in a couple of issues/pull requests in octokit lately regarding `Webhooks` typing inference. What I found is these comments: 
- https://github.com/octokit/app.js/pull/201#discussion_r576445234
- https://github.com/octokit/app.js/pull/201#issuecomment-781013558 

Do you think [this](https://devblogs.microsoft.com/typescript/announcing-typescript-4-2/#smarter-type-alias-preservation) new improvement in TS 4.2 could help in Octokit type inference problem with `Webhooks`?

Here I would like some TS expertise from @wolfy1339 and @G-Rath

**Comment by gr2m** [comment]
Very cool! I'm excited to see how this change impacts types for `@octokit/*` modules!

You could try reverting https://github.com/octokit/app.js/pull/201 and installing TypeScript 4.2 locally, then see what type is shown in VS Code?

**Reply by oscard0m** [reply]
Seems like there is no difference? 😢  (I'm doing mouse hover for `octokit.webhooks` property, that would be the check right?)

https://codesandbox.io/s/nice-lewin-i76jc?file=/src/index.ts

🆚 

https://codesandbox.io/s/octokit-app-1021-typescript-422-eypu6?file=/src/index.ts

**Reply by wolfy1339** [reply]
I get the same result locally when compiling the source TS files into definition files using 4.2

**Reply by wolfy1339** [reply]
I don't know if it's me, but octokit/app.js#201 doesn't seem to have an effect at all for me when using the App#webhooks method.

**Reply by oscard0m** [reply]
> I don't know if it's me, but [octokit/app.js#201](https://github.com/octokit/app.js/pull/201) doesn't seem to have an effect at all for me when using the App#webhooks method.

By effect what do you mean, types when you hover are the same ones?

**Reply by wolfy1339** [reply]
Yes, exactly

**Reply by wolfy1339** [reply]
I'm using VSCode on Ubuntu 20.10 and this is the output of the About menu
```
Version: 1.53.2
Commit: 622cb03f7e070a9670c94bae1a45d78d7181fbd4
Date: 2021-02-11T11:48:44.518Z
Electron: 11.2.1
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Linux x64 5.8.0-43-generic
```
No VSCode updates are available. Same results with using TS 4.2 as with 4.1

**Reply by oscard0m** [reply]
@wolfy1339 Could you try running `tsc` on it? Checking the [issue](https://github.com/octokit/app.js/issues/200) seems to be happening when running `tsc`.

**Reply by wolfy1339** [reply]
Running `tsc` doesn't seem to do anything to fix it. It just transpiles the TS source into JS and doesn't do anything with the type information

**Reply by oscard0m** [reply]
~This is the full command reported in the issue, maybe you need to run the full command?~

~`node_modules/typescript/bin/tsc --outDir pkg/dist-src/ -d --declarationDir pkg/dist-types/ --project tsconfig.json --target es2020 --module esnext --noEmit false --sourceMap false
src/webhooks.ts:8:17`~

Check run with details here: https://github.com/octokit/app.js/runs/1858152718?check_suite_focus=true#step:5:14

`npm run build` in the project or `pika build` downgrading the commit and trying TS4.2 vs TS4.1?

**Reply by wolfy1339** [reply]
That's exactly the process I followed when I tested

**Reply by oscard0m** [reply]
Strange. What we could do is, first make sure TS4.2 is used everywhere it Octokit and, from there opening a PR 'reverting' this types and with Pika's generated package play with it and see if we are able to reproduce the error again.

**Comment by G-Rath** [comment]
4.2 has actually been released a few hours ago, so we can upgrade and see what happens.
